File: introduction/ITP1_11_B-Dice II.rb
class Dice

  def initialize(labels)
    @labels = labels
    @patterns = []
    7.times do |i|
      roll(i<4?'N':'W')
      next if i == 5
      @patterns.concat(all(@labels))
    end
  end

  def roll(dir)
    a, b, c, d, e, f = @labels
    case dir
    when 'N' then @labels = [b, f, c, d, a, e]
    when 'S' then @labels = [e, a, c, d, f, b]
    when 'E' then @labels = [d, b, a, f, e, c]
    when 'W' then @labels = [c, b, f, a, e, d]
    end
  end

  def top
    @labels[0]
  end

  def all(l)
    return [[l[0], l[1], l[2], l[3], l[4], l[5]], [l[0], l[2], l[4], l[1], l[3], l[5]], [l[0], l[4], l[3], l[2], l[1], l[5]], [l[0], l[3], l[1], l[4], l[2], l[5]]]
  end

  def patterns
    @patterns
  end

end

dice = Dice.new(gets.split.map(&:to_i))
gets.chomp.to_i.times do
  t, f = gets.chomp.split.map(&:to_i)
  dice.patterns.each{|p|
    if p.slice(0..1) == [t, f] then
      puts p[2]
      break
    end
  }
end

File: introduction/ITP1_11_B-Dice II.scala
case class Dice(var label:Array[String]) {
  def rotate(dir:Char) =
    dir match {
      case 'N' => Array(1,5,2,3,0,4)
      case 'S' => Array(4,0,2,3,5,1)
      case 'E' => Array(3,1,0,5,4,2)
      case 'W' => Array(2,1,5,0,4,3)
    }
  def roll(dirs:String) =
    label = dirs.foldLeft(Array(0,1,2,3,4,5)) { (x,y) => rotate(y).map(x) }.map(label)
  var patterns = List[Array[String]]()
  for(i <- 0 to 7 if i != 5) {
    roll(if(i<4)"N" else "W");
    patterns :+= label
  }
  print(patterns.size)
}
object Main extends App {
  val dice = Dice(readLine.split(" "))
  val n = readInt
  (0 until n).foreach { i =>
    val l = readLine.split(" ")
    println(l.mkString(" "))
  }
}
